refactor skinning and animating framework

shrinktofit opened the pull request:
* add onLoad() callback when asset is loaded;
* discard class SkeletonInstance, refactor skinning and animating framework
* discard interface AnimationResource and it's relevant classes, directly store animation clip data into AnimationClip asset
* add getNodeByPath() API to class Node

shrinktofit committed:
* add onLoad() callback when asset is loaded;
* discard class SkeletonInstance, refactor skinning and animating framework
* discard interface AnimationResource and it's relevant classes, directly store animation clip data into AnimationClip asset
* add getNodeByPath() API to class Node

shrinktofit committed:
uuid-loader.js

shrinktofit committed:
bug fix

pandamicro commented on cocos/load-pipeline/uuid-loader.js:
```diff
@@ -288,7 +288,12 @@ export function loadUuid (item, callback) {
-    loadDepends(this.pipeline, item, asset, depends, callback);
+    loadDepends(this.pipeline, item, asset, depends, (...args)=> {
+        if (asset.onLoad) {
+            asset.onLoad();
```
@shrinktofit 推动一下跟 Jare 的讨论，确定一下这里应该要使用的 API

https://github.com/cocos-creator/3d-tasks/issues/127

shrinktofit committed:
Merge branch '3d' into leslie2

dabingnn commented on cocos/scene-graph/base-node.js:
```diff
@@ -459,6 +459,27 @@ export default class BaseNode extends CCObject {
+    /**
+     * !#en Returns a child from the container given its path.
+     * !#zh 通过路径获取节点的子节点。
+     * @method getChildByName
```
typo

dabingnn commented on cocos/scene-graph/base-node.js:
```diff
@@ -459,6 +459,27 @@ export default class BaseNode extends CCObject {
+        const segments = path.split('/');
+        let lastNode = this;
+        for (const segment of segments) {
+            lastNode = lastNode.children.find(childNode => childNode.name === segment);
```
how to deal with wrong path?

shrinktofit replied:
Wrong path cause this function simply return null.

shrinktofit committed:
Make camera-component is available in previewing;
Remove onLoad of mesh asset class, instead do its content in model-component and skinning-model-component's _updateModels method.

dabingnn commented on cocos/3d/framework/skinning-model-component.js:
```diff
@@ -80,40 +93,31 @@ export default class SkinningModelComponent extends ModelComponent {
+    }
+
+    onEnable() {
+        this._resetTarget();
```
should we recreate `skinningTarget` every time when enabled? could we create it when it is loaded?

shrinktofit replied:
I've done it.

shrinktofit committed:
Merge pull request #1 from shrinktofit/leslie

Improve

dabingnn commented on cocos/3d/assets/animation-clip.js:
```diff
@@ -33,387 +33,366 @@ import { clamp } from "../../core/vmath/utils";
+    }
+
+    _saveDefaultTRS() {
+        this._defaultPositions = new Array(this._nodes.length);
```
there are two ways in this code for default position, scale and rotation, which one is better?

shrinktofit replied:
This function currently isn't used. I have commented it out.

shrinktofit committed:
typo

shrinktofit committed:
Merge pull request #2 from shrinktofit/leslie

typo

shrinktofit committed:
move _resetTarget() to onLoad

shrinktofit committed:
Merge pull request #3 from shrinktofit/leslie

move _resetTarget() to onLoad
